### ISSUE TITLE
redundant_clone does not fire when `String::push_str` is called

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust-clippy#9988]
### Summary

Calling `String::push_str` on a `String` before needlessly cloning it with `ToString::to_string` causes the lint to not fire.

### Lint Name

redundant_clone

### Reproducer

I tried this code:

```rust
#![deny(clippy::all)]

fn demo() -> String {
    let mut result = String::new();
    result.push_str("abc");
    result.to_string()
}
```

I expected to see this happen:

```
error: redundant clone
 --> src/lib.rs:6:11
  |
6 |     result.to_string()
  |           ^^^^^^^^^^^^ help: remove this
  |
note: this value is dropped without further use
 --> src/lib.rs:6:5
  |
6 |     result.to_string()
  |     ^^^^^^
```

Instead, this happened:

No warnings.

### Version

```text
cargo 1.67.0-nightly (ba607b23d 2022-11-22)
release: 1.67.0-nightly
commit-hash: ba607b23db8398723d659249d9abf5536bc322e5
commit-date: 2022-11-22
host: aarch64-apple-darwin
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.84.0 (sys:0.4.59+curl-7.86.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
os: Mac OS 13.0.1 [64-bit]
```
